### PR TITLE
[nnfw] Interpret quant8 asymm as uint8 @open sesame 1/6 13:47

### PR DESCRIPTION
**Pull request by kparichay**
Interpret quant8 asymm as uint8 (after discussion with @chunseoklee)
Add unittest with object classification

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Comment by dongju-chae** [comment]
It seems that your PR #1985 includes `tests/test_models/data/orange.raw`.

File: tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp [hunk 1]
```diff
@@ -250,6 +255,163 @@ TEST (nnstreamer_nnfw_runtime_raw_functions, invoke)
...
+	}
+
+  manifest_file = g_build_filename (root_path, "tests", "test_models", "models",
+      "metadata", "MANIFEST", NULL);
+	status = g_file_test (manifest_file, G_FILE_TEST_EXISTS);
+  if (status == FALSE) {
+    g_free (model_file);
+    g_free (manifest_file);
+    ASSERT_EQ (status, TRUE);
+	}
```
Comment (dongju-chae) [inline tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: needs indent fix.

**Comment by kparichay** [comment]
> It seems that your PR #1985 includes `tests/test_models/data/orange.raw`.

Yes, they both are the same file. Merging one of the PR will automatically remove the other file I think as they will match exactly.

File: tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp (hunk 1 above)
Comment (kparichay) [reply tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: Updated.

File: tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp [hunk 2]
```diff
@@ -250,6 +255,163 @@ TEST (nnstreamer_nnfw_runtime_raw_functions, invoke)
...
+  manifest_file = g_build_filename (root_path, "tests", "test_models", "models",
+      "metadata", "MANIFEST", NULL);
+  status = g_file_test (manifest_file, G_FILE_TEST_EXISTS);
+  if (status == FALSE) {
+    g_free (model_file);
+    g_free (manifest_file);
+    ASSERT_EQ (status, TRUE);
+  }
+
+  replace_command = g_strdup_printf("sed -i '/%s/c\\\"models\" : [ \"%s\" ],' %s",
```
Comment (gichan-jang) [inline tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: Is this line to use one MANIFEST file? 
That means... When we want to use multiple models, we have to use this way?
Maybe it related to issue #1981.

Comment (kparichay) [reply tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: This line modifies the MANIFEST file to tell the NNFW to use the desired model.
Yes, its related to #1981. The method used in the unittest is not the best way (this overwrites the complete line and assumes a specific format of the line previously), but works for this unittest. A better way would be to parse the MANIFEST file and then adjust it properly, which will be discussed in #1981.

File: tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp [hunk 3]
```diff
@@ -250,6 +255,163 @@ TEST (nnstreamer_nnfw_runtime_raw_functions, invoke)
...
+  g_free (data_file);
+  g_free (output.data);
+  g_free (input.data);
+  g_free (model_file);
+  sp->close (&prop, &data);
+
+  replace_command = g_strdup_printf("sed -i '/%s/c\\\"models\" : [ \"%s\" ],' %s",
+      new_model, orig_model, manifest_file);
+  ret = system(replace_command);
+  g_free(replace_command);
```
Comment (dongju-chae) [inline tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: It's minor comment: need a space between a function name and its arguments for uniformity.

File: tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp [hunk 4]
```diff
@@ -250,6 +255,163 @@ TEST (nnstreamer_nnfw_runtime_raw_functions, invoke)
...
+
+  output.type = res.info[0].type;
+  output.size = gst_tensor_info_get_size (&res.info[0]);
+
+  input.data = g_malloc(input.size);
+  output.data = g_malloc(output.size);
+
+  data_file = g_build_filename (root_path, "tests", "test_models", "data",
+      "orange.raw", NULL);
+  fd = open(data_file, O_RDONLY);
```
Comment (jaeyun-jung) [inline tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: ~line 392 recommended to use g_file_get_contents()

File: tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp (hunk 3 above)
Comment (kparichay) [reply tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: updated.

File: tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp (hunk 4 above)
Comment (kparichay) [reply tests/tizen_nnfw_runtime/unittest_tizen_nnfw_runtime_raw.cpp]: updated.

File: ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c [hunk 5]
```diff
@@ -281,12 +285,11 @@ nnfw_tensor_info_copy (const nnfw_tensorinfo * nnfw_info,
   if ((status = nnfw_tensor_type_to_gst (nnfw_info->dtype, &gst_info->type)) < 0)
     return status;
 
-  /** reverse the order of dimension */
-  for (idx = 0; idx < nnfw_info->rank; idx ++)
-    gst_info->dimension[NNS_TENSOR_RANK_LIMIT - idx - 1] = nnfw_info->dims[idx];
+  for (idx = nnfw_info->rank-1; idx >= 0; idx--)
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c]: check the style (gst-indent)

Comment (myungjoo) [reply ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c]: Can you sure that idx won't go beyond 3?

File: ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c [hunk 6]
```diff
@@ -393,6 +396,10 @@ nnfw_tensor_type_from_gst (const tensor_type type, NNFW_TYPE * nnfw_type)
     case _NNS_INT32:
       *nnfw_type = NNFW_TYPE_TENSOR_INT32;
       break;
+    case _NNS_UINT8:
+      /** @todo: update this to NNFW_TYPE_TENSOR_UINT8 type once nnfw is updated */
+      *nnfw_type = NNFW_TYPE_TENSOR_QUANT8_ASYMM;
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c]: When they update QUANT8_ASYMM to UINT8 in the future, they will get build errors from OBS/Dashboard. They will be required to do "Group-SR" at that time. Simply be notified for this issue; you will be required to do group-SR with nnfw developers at the same time.

File: ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c (hunk 5 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c]: > check the style (gst-indent)

Updated

Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c]: > Can you sure that idx won't go beyond 3?

Yes, line 281 does this check on the rank of nnfw_info.

File: ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c (hunk 6 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c]: I will do Group-SR.
Note: They will not update QUANT8_ASYMM to UINT8. They will add an extra data type UINT8.

**Pull request merged**